include/irr/asset/IVirtualTexture.h

```diff
@@ -0,0 +1,750 @@
+#ifndef __IRR_I_VIRTUAL_TEXTURE_H_INCLUDED__
+#define __IRR_I_VIRTUAL_TEXTURE_H_INCLUDED__
+
+#include "irr/asset/format/EFormat.h"
+#include "irr/core/alloc/GeneralpurposeAddressAllocator.h"
+#include "irr/core/alloc/PoolAddressAllocator.h"
+#include "irr/core/math/morton.h"
+#include "irr/core/alloc/address_allocator_traits.h"
+#include "irr/core/memory/memory.h"
+#include "irr/asset/filters/CPaddedCopyImageFilter.h"
+#include "irr/asset/filters/CFillImageFilter.h"
+
+namespace irr {
+namespace asset
+{
+
+template <typename image_view_t>
+class IVirtualTexture : public core::IReferenceCounted
+{
+protected:
+    //! SPhysPgOffset is what is stored in texels of page table!
+    struct SPhysPgOffset
+    {
+        _IRR_STATIC_INLINE_CONSTEXPR uint32_t invalid_addr = 0xffffu;
+
+        SPhysPgOffset(uint32_t _addr) : addr(_addr) {}
+
+        //upper 16 bits are used for address of mip-tail page
+        uint32_t addr;
+    };
+
+    uint32_t neededPageCountForSide(uint32_t _sideExtent, uint32_t _level)
+    {
+        return (std::max(_sideExtent>>_level, 1u)+m_pgSzxy-1u) / m_pgSzxy;
+    }
+
+private:
+    uint32_t computeSquareSz(const IImage* _img, const ICPUImage::SSubresourceRange& _subres)
+    {
+        auto extent = _img->getCreationParameters().extent;
+
+        const uint32_t w = neededPageCountForSide(extent.width, _subres.baseMipLevel);
+        const uint32_t h = neededPageCountForSide(extent.height, _subres.baseMipLevel);
+
+        return core::roundUpToPoT(std::max(w, h));
+    }
+
+public:
+    _IRR_STATIC_INLINE_CONSTEXPR uint32_t MAX_PHYSICAL_PAGE_SIZE_LOG2 = 9u;
```

Comment:
why is this a constexpr?

we should support 128,256,512 and 1024 pages.. we shouldn't hardcode page size.

Reply:
it's not page size, it's page size limit (limit log2 is 9 so max page size is 512). IIRC our biggest supportable page size was supposed to be 512, dont remember why, i may remember badly

Reply:
ok, thats fine then.